File: leetcode-04/solutions/cpp/0018.cpp
class Solution {
 public:
  vector<vector<int>> fourSum(vector<int>& nums, int target) {
    if (nums.size() < 4) return {};

    vector<vector<int>> ans;

    sort(nums.begin(), nums.end());

    for (int i = 0; i < nums.size() - 3; i++)
      if (i == 0 || nums[i] != nums[i - 1])
        for (int j = i + 1; j < nums.size() - 2; j++)
          if (j == i + 1 || nums[j] != nums[j - 1]) {
            int l = j + 1;
            int r = nums.size() - 1;
            int newTarget = target - nums[i] - nums[j];
            while (l < r) {
              if (nums[l] + nums[r] == newTarget) {
                ans.push_back({nums[i], nums[j], nums[l], nums[r]});
                while (l < r && nums[l] == nums[l + 1]) l++;
                while (l < r && nums[r] == nums[r - 1]) r--;
                l++;
                r--;
              } else if (nums[l] + nums[r] < newTarget) {
                l++;
              } else {
                r--;
              }
            }
          }

    return ans;
  }
};

File: leetcode-04/solutions/python3/0018.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4:
            return []

        ans = []

        nums.sort()

        for i in range(len(nums) - 3):
            if i == 0 or nums[i] != nums[i - 1]:
                for j in range(i + 1, len(nums) - 2):
                    if j == i + 1 or nums[j] != nums[j - 1]:
                        l = j + 1
                        r = len(nums) - 1
                        newTarget = target - nums[i] - nums[j]
                        while l < r:
                            if nums[l] + nums[r] == newTarget:
                                ans.append(
                                    [nums[i], nums[j], nums[l], nums[r]])
                                while l < r and nums[l] == nums[l + 1]:
                                    l += 1
                                while l < r and nums[r] == nums[r - 1]:
                                    r -= 1
                                l += 1
                                r -= 1
                            elif nums[l] + nums[r] < newTarget:
                                l += 1
                            else:
                                r -= 1

        return ans
